src/celeritas/optical/Transporter.hh

```diff
@@ -0,0 +1,80 @@
+//------------------------------- -*- C++ -*- -------------------------------//
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/Transporter.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <memory>
+
+#include "corecel/Types.hh"
+#include "corecel/math/NumericLimits.hh"
+#include "corecel/sys/ActionGroups.hh"
+#include "celeritas/Types.hh"
+
+#include "CoreState.hh"
+
+namespace celeritas
+{
+template<class P, template<MemSpace M> class S>
+class ActionGroups;
+
+namespace optical
+{
+class CoreParams;
+template<MemSpace M>
+class CoreState;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Transport all pending optical tracks to completion.
+ */
+class Transporter
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using CoreStateHost = CoreState<MemSpace::host>;
+    using CoreStateDevice = CoreState<MemSpace::device>;
+    using SPConstParams = std::shared_ptr<CoreParams const>;
+    //!@}
+
+    struct Input
+    {
+        SPConstParams params;
+        size_type max_step_iters{numeric_limits<size_type>::max()};
+
+        //! True if all input is assigned and valid
+        explicit operator bool() const { return static_cast<bool>(params); }
+    };
+
+  public:
+    // Construct with problem parameters and setup options
+    explicit Transporter(Input);
+
+    // Transport all pending optical tracks on the host
+    void operator()(CoreParams const&, CoreStateHost&) const;
+
+    // Transport all pending optical tracks on the device
+    void operator()(CoreParams const&, CoreStateDevice&) const;
```

Comment:
Since we're constructing with SP<CoreParams> maybe we should just call it with the state?